File: python/medium/2962_count_subarrays_where_max_element_appears_at_least_k_times.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left, freq, answer = 0, 0, 0
        max_value = max(nums)
        for right in range(n):
            if nums[right] == max_value:
                freq += 1
            while freq >= k:
                answer += n - right
                if nums[left] == max_value:
                    freq -= 1
                left += 1
        return answer

File: rust/medium/2962_count_subarrays_where_max_element_appears_at_least_k_times.rs
impl Solution {
    pub fn count_subarrays(nums: Vec<i32>, k: i32) -> i64 {
        let (mut left, mut freq, mut answer) = (0, 0, 0);
        let max = nums.iter().max().unwrap();
        for right in 0..nums.len() {
            let entry = nums[right];
            if &entry == max {
                freq += 1;
            }
            while freq >= k {
                answer += (nums.len() - right) as i64;
                if &nums[left] == max {
                    freq -= 1;
                }
                left += 1;
            }
        }
        answer
    }
}
